### DISCUSSION TITLE
Bilayers v0.1

### DISCUSSION THREAD

**Discussion by gnodar01** [bilayer-containers/bilayers#5] [Ideas]
Our goal for bilayer is to work up to whats shown in the following figure:

![COBA_R24_proposal](https://github.com/bilayer-containers/bilayers/assets/8888732/d224e5a7-3a50-4c0a-921d-a553dce472d3)

We'll take it one step at a time though. For version `0.1`, @rajavishah and I discussed concentrating on an MVP. The idea is to build off the docker gradio demo that Callum built. We [have a branch with some rough notes](https://github.com/COBA-NIH/docker_gradio_demo/tree/biapy_notes) in that repo.

For `v0.1`, we will start by only creating enough to programmatically generate a replica of the threshold gradio app. We'll add other interfaces afterwards.

The MVP will have the following starting elements:
* a super simple (for now) config/spec file (maybe in json, or yaml, or toml, or ...) definition
* a docker file which installs all the gradio dependencies

We'll then need components that are specific to the algorithm, in this case simple threshold:
* populated config/spec file
  * input & output UI types (for now just `filepath`, `choice` / `radio` and `number` is needed; we will expand UI types later)
* A second dockerfile which `FROM`s the base gradio docker file, and then installs the necessary dependencies  (`scikit-image` and `scipy` in this case)
* a python file providing the "execute" function (i.e. the [example function](https://github.com/COBA-NIH/docker_gradio_demo/blob/biapy_notes/app.py#L22))

Finally we'll need a build process. Initially the plan is to simply have a python script which parses the config/spec file and:
* programmatically generates the Gradio interface
  * corresponding to [these lines](https://github.com/COBA-NIH/docker_gradio_demo/blob/biapy_notes/app.py#L9-L20), followed by the `execute` function from above, and then [these lines](https://github.com/COBA-NIH/docker_gradio_demo/blob/biapy_notes/app.py#L75-L83)
* a third dockerfile which `FROM`s the previous one above, and copies in the programmatically generated file above, and sets its execution as the `ENTRYPOINT`

We will also need to come up with a folder structure that organizes all this

**Comment by gnodar01** [comment]
@bethac07 Did I miss something obvious?

@rajavishah Do you think this properly captures what we discussed? If so, once we come to a consensus that this is a good path forward, let's follow the plan in discussion #2 and transfer this to an issue, and create the corresponding milestone and project (I think an `mvp` project would be good, with a `v0.1` milestone). Then create the corresponding branch and file a new draft PR.

**Reply by bethac07** [reply]
I don't 100% get the "middle" Docker - if it's just pip installing stuff into the gradio docker, do we need that? Is it just to do dependency resolution? I think in my head gradio versions, etc were just handled at the spec level (ie, `bilayers-gradio-v1` tries to install `gradio==somever, some_other_package=some_other_ver` and grabs `stock-gradio-file-v1-url`), etc), but I wasn't there for the discussions you folks have been having so I fully believe in me having missed some other reason why it's valuable to have.

**Reply by gnodar01** [reply]
I may be overthinking it, but I tried hard to underthink my overthinking and failed to come up with anything sufficient.

The thought process was that the spec file specifying `bilayer-gradio-v1` has to eventually translate to: `RUN python -m pip install gradio==3.25.0 gradio_client==0.1.3 huggingface-hub==0.14.0 pydantic==1.10.7`.

The logic behind that can go in the build script, but its probably easier and more clear to just have a `gradio_v1_dockerfile`. We would have to do this once for every interface (and its version) that we care about.

We will then have a list of algorithms we care about, and each will have a unique set of dependencies. For the simple case of threshold its just `scikit-image` and `scipy`, so we could in theory have just a `requirements.txt`. That however doesn't work for more complicated algorithm dependencies. For example we may have to do some `apt-install` stuff as well, or compile something, or set env flags specific to that algorithm. So its easier just to pack it all into its own dockerfile as well. We would have to do this once for every algorith (and its version) that we care about.

The third dockerfile is just copying in the final generated `app.py` (e.g. `ADD app.py /workspace/`), and setting the port (e.g. `EXPOSE 8000` and entrypoint (e.g. `CMD [ "python3" , "/workspace/app.py" ]`). It does no installations. Technically this can just be handled by `docker run` CLI flags, so it is a bit extra, but its just the generated artifact from running the build. We don't write it ourselves, and it doesn't actually go in the repository contents. We can use [multi-stage](https://docs.docker.com/build/building/multi-stage/) builds and combine the above two dockerfiles, matrixed out across base dockerfiles (gradio, jupyter, ...), and algorithm dockerfiles (threshold, stardist, ...).

The result is what the user gets (or is the thing we run `docker build` on to generate the docker image, which is what the user gets)

**Reply by bethac07** [reply]
Ah, ok, here's the thing I think where you and I differed - wherever possible, I DON'T want to make the "wrappee" Dockerfile. I want to use a Docker that already exists, because I am lazy. So idea being
- Dev makes a cool DNN, they make it into a Docker (or we do, and yes, this part may be annoying, but it seems like the field is kind of going this way in terms of people making them)
- We wrap it in a second Docker that brings in pip-installable dependencies and copies in an app.py file that was made in a script.
- There is no third Docker

Otherwise, if we start from the interfaces and then add the network, we do 1 installation of the interface and (n networks) installations of the network, which is likely the harder install. If we start from the network and add the interfaces, we do one installation of the network and then (n interfaces) version of the interfaces, which is hopefully the easier install.  Does that make sense?

**Reply by bethac07** [reply]
Documenting before I forget - we had a quick call on this, and the discrepancy is that I'm still living pre-2020 and didn't know about [multistage builds.](https://docs.docker.com/build/building/multi-stage/) . We will therefore try that first, and then if there are too many complexities with having to know things about the wrappee network, we will go to moving stuff to the CI

**Reply by gnodar01** [reply]
@rajavishah Beth and I discussed over a call yesterday.

First, Beth is correct about the order. It definitely makes way more sense for the NN/Algo to be the thing that gets wrapped. Also if the NN author already publishes a docker image, all the better.

Second, Beth's idea around the wrapper docker was to have it contain variables plus some `sed` calls from the CI/CD, to generate the dockerfile responsible for both installing the interface dependencies AND configuring any additional things like copying in the `app.py` file, exposing `ports`, etc. In other words take a template dockerfile and transform it into the final dockerfile that defines the NN/Algo + interface + entrypoint. Whereas my idea is to have a dockerfile that only defines the interface installations. For instance you might have something like:

Algorithm dockerfile (defined by us, or published by the algo author):

```
FROM python:3.9-slim

RUN python -m pip install scipy==1.9.1 scikit-image==0.20.0

.. some other stuff we don't care about
```

Template dockerfile for gradio interface

```
ARG BASE_TOOL
ARG TOOL_VERSION

FROM ${BASE_TOOL}:${TOOL_VERSION}

WORKDIR /bilayer_src

RUN python -m pip install gradio==3.25.0 gradio_client==0.1.3 huggingface-hub==0.14.0 pydantic==1.10.7

ARG PORT=8000
ARG ENTRY_SCRIPT="app.py"

ADD ${ENTRY_SCRIPT} ./

EXPOSE ${PORT}

CMD ["python3", "/workspace/${ENTRY_SCRIPT}"]
```

Then in the build process (CI/CD, etc) `docker build --build-arg BASE_TOOL="simple_threshold" TOOL_VERSION="1.0" .`

@bethac07 Please do let me know if I have misunderstood what you were suggesting there.

My idea, was to have an interface docker for just the dependencies, and then use multi-stage build to `FROM` both the NN/Algo and Interface dockers, set the entrypoint, etc. We ended the call with this approach in mind, but on further reflection I think it ends up being more complicated. Since we have little control over the base that the NN/Algo builds on, we don't know in advance what base to build on for the interface docker (e.g. alpine vs debian vs ...). So I think Beth's idea is the way to go (unless @rajavishah has other thoughts).

I'll circle back to it later to validate/invalidate whether its a viable option or not.

**Reply by bethac07** [reply]
Well summarized!
